File: Stream_Kmeans_Scala/k-means.scala
/* 
============================================
=== FIT5202 Data Processing for Big Data ===
======= Assignment II - PoC1 K-Means =======
============================================

===== GROUP 3: =====
Arno Angerer, 29909163
Sneha Ravikumar, 29391911
Marek Schumann, 29971683
Josefine Tusindfryd, 29876893
Richard Osakwe, 29466695 


========================================
============== DESCRIPTION =============
========================================

This file contains our proof of concept implementation for
the K-Means Streaming Algorithm. This implementation is heavily
inspired by the StreamingKMeans sample implementation found here:
https://github.com/apache/spark/blob/branch-1.2/examples/src/main
/scala/org/apache/spark/examples/mllib/StreamingKMeansExample.scala

========================================
================ USAGE =================
========================================

To execute the algorithm from within the BigVM, this k-means.scala
script should be placed in a folder together with the k-means.py script.
Also, the two subdirectories, './testDir' and './trainingDir', should be present.

With this environment in place, start the K-Means Streaming algorithm via
$ cat k-means.scala | spark-shell

This starts the Spark streaming algorithm. Our program is now actively monitoring the
'./trainingDir' directory for any new data files. If new data is found it will 
automatically be parsed and used to update the cluster centroids accordingly.

Then, from another terminal window in the same directory, execute the provided
python script to start generating a sample data flow:
$ python k-means.py 

This script will generate random data files with in 2 second intervals so that the
clustering process can be observed in real-time. As we add text files to `trainingDir` 
the clusters will continuously update. Then, when we add the file to `testDir` we will
obtain the predicted labels using the most up-to-date model.
*/


// Import Relevant Libraries
import org.apache.spark.SparkConf
import org.apache.spark.mllib.clustering.StreamingKMeans
import org.apache.spark.mllib.linalg.Vectors
import org.apache.spark.mllib.regression.LabeledPoint
import org.apache.spark.streaming.{Seconds, StreamingContext}

// Define dynamic Streaming Object
object StreamingKMeansPoC {

  def main(args: Array[String]) {
    if (args.length != 4) {
      System.err.println(
        "Error: Wrong number of arguments provided." +
        "Please provide the following arguments:" +
          "<trainingDir> <testDir> <interval> <number_of_clusters>")
      System.exit(1)
    }

    val conf = new SparkConf().setMaster("local").setAppName("StreamingKMeansPoC")
    
    // Initialise Spark Streaming Context
    val ssc = new StreamingContext(conf, Seconds(args(2).toLong))

    // Set up directory monitoring streams and appropriate parsers
    val trainingData = ssc.textFileStream(args(0)).map(Vectors.parse)
    val testData = ssc.textFileStream(args(1)).map(LabeledPoint.parse)

    // Create StreamingKMeans model object
    val model = new StreamingKMeans()
      .setDecayFactor(1.0) // no decay
      .setK(args(3).toInt) // number of clusters
      .setRandomCenters(2, 0.0) // initialise random centroids (2d)

    // Specify incoming stream to train model on
    model.trainOn(trainingData) 

    // Use the model to make predictions and print results (label + cluster)
    model.predictOnValues(testData.map(lp => (lp.label.toInt, lp.features))).print()
    
    // Store latest model for inspection and debugging
    val latestModel = model.latestModel().clusterCenters

    ssc.start()
    ssc.awaitTermination()
  }
}

// Stop previously running Spark Context
sc.stop()

// Run K-Means with desired properties i.e. 
// 3s batch interval and 3 clusters
StreamingKMeansPoC.main(Array("trainingDir","testDir","3","3"))

File: Stream_Kmeans_Scala/k-means.py
# ============================================
# === FIT5202 Data Processing for Big Data ===
# ======= Assignment II - PoC1 K-Means =======
# ============ Auxilary Script ===============
# ============================================
# 
# ===== GROUP 3: =====
# Arno Angerer, 29909163
# Sneha Ravikumar, 29391911
# Marek Schumann, 29971683
# Josefine Tusindfryd, 29876893
# Richard Osakwe, 29466695 

# ========================================
# ============== DESCRIPTION =============
# ========================================
# 
# This file contains the auxilary python script that
# creates a stream of random datapoints to be consumed
# by our K-Means clustering application.
# See the `k-means.scala` file or the accompanying
# documentation for usage instructions.


# Import relevant libraries
import os
import random
import time

# Clean up data files from last run
os.system("rm ./testDir/* 2> /dev/null")
os.system("rm ./trainingDir/* 2> /dev/null")


# Create random points belonging to one of three clusters:
def create_random_points(n = 10,start_idx = 1):
    random_points = [(random.randint(0,12),random.randint(0,18)) for x in range(n//3)]
    random_points.extend([(random.randint(20,34),random.randint(30,62)) for x in range(n//3)])
    random_points.extend([(random.randint(60,95),random.randint(70,100)) for x in range(n//3)])
    random_points = (list(enumerate(random_points,start_idx)))
    next_idx = start_idx + len(random_points)
    return (random_points,next_idx)


# Define function to write train and test files
def create_files(list_of_points, iteration):
    # Create train file
    train_file_handle = open("./trainingDir/train" + str(iteration + 1) + ".txt", 'w')
    for point in list_of_points:
        train_file_handle.write("[" + str(point[1][0]) + "," + str(point[1][1]) + ']\n')
    train_file_handle.close()
    
    # Create test file
    test_file_handle = open("./testDir/test" + str(iteration + 1) + ".txt", 'w')
    for point in list_of_points:
        test_file_handle.write("(" + str(point[0]) + ",[" + str(point[1][0]) + "," + str(point[1][1]) + "])\n")
    test_file_handle.close()

# Create 'stream' of incoming data files
idx = 1 # set starting index
number_of_files = 10
datapoints_per_file = 10
interval_of_arrival = 2
for batch in range(number_of_files):
    data, idx = create_random_points(datapoints_per_file,idx)
    create_files(data,batch)
    time.sleep(interval_of_arrival)